arangod/Aql/OptimizerRulesCluster.cpp

```diff
@@ -439,6 +434,111 @@ bool substituteClusterSingleDocumentOperationsNoIndex(
   return modified;
 }
 
+bool substituteClusterMultipleDocumentInsertOperations(
+    Optimizer* opt, ExecutionPlan* plan, OptimizerRule const& rule) {
+  containers::SmallVector<ExecutionNode*, 8> nodes;
+  plan->findNodesOfType(nodes, {EN::INSERT}, false);
+
+  if (nodes.size() != 1) {
+    return false;
+  }
+
+  auto* node = nodes[0];
+  auto* dep = node->getFirstDependency();
+  if (dep == nullptr || dep->getType() != EN::ENUMERATE_LIST) {
+    return false;
+  }
+
+  if (!::depIsSingletonOrConstCalc(dep)) {
+    return false;
+  }
+
+  bool modified = false;
+  auto mod = ExecutionNode::castTo<InsertNode*>(node);
+
+  Variable const* oldVariable = mod->getOutVariableOld();
+  if (oldVariable != nullptr && mod->isVarUsedLater(oldVariable)) {
+    // using RETURN OLD cannot use optimization
+    return false;
+  }
+
+  Variable const* newVariable = mod->getOutVariableNew();
+  if (newVariable != nullptr && mod->isVarUsedLater(newVariable)) {
+    // using RETURN NEW. cannot use optimization
+    return false;
+  }
+
+  auto* enumerateNode = ExecutionNode::castTo<EnumerateListNode const*>(dep);
+  if (enumerateNode->outVariable() != mod->inVariable()) {
+    return false;
+  }
+
+  if (enumerateNode->isInInnerLoop()) {
+    // FOR ... INSERT is contained in inner loop. cannot use optimization
+    return false;
+  }
+
+  {
+    VarSet usedVars;
+    node = node->getFirstParent();
+    while (node) {
```

Comment:
I will remove it for now, but we intend to add a patch that supports RETURN, including OLD and NEW, after transaction is el cheapo and created in Query.cpp, in the plan, instead of creating it inside the Executor